docs/dev/relay_pass_infra.rst

```diff
@@ -0,0 +1,615 @@
+a normal form conversion and lambda lifting, etc, fall in this set. At this
+level, uses can even add and/or delete functions in a module.
+
+.. code:: c++
+
+    class ModulePassNode : PassNode {
+      PassInfo pass_info;
+      runtime::TypedPackedFunc<Module(Module, PassContext)> pass_func;
+      Module operator()(const Module& mod, const PassContext& pass_ctx) const final;
+      // Other members/methods are omitted
+    };
+
+``pass_info`` maintains the information needed by a module-level pass.
+``pass_func`` sketches the real optimization. For example, we may need to
+perform dead code elimination on the module. We could implement the algorithm in
+the ``pass_func`` and let it run on a module. It will then remove the dead code
+including the unused functions in the module. Note that this field is designed
+as a packed function, which enables the implementation of the optimization in
+both C++ and Python.
+
+Function-level Passes
+^^^^^^^^^^^^^^^^^^^^^
+
+Function-level passes are used to implement various intra function level
+optimizations for a given Relay module. It fetches one function at a time from
+the function list of a module for optimization and yields a rewritten Relay
+function. Most of Relay passes can be classified into this category, such as
+common subexpression elimination and inference simplification, etc.
+
+Note that the scope of passes at this level is a Relay function. Therefore, we
+cannot add or delete a function through these passes as they are not aware of
+the global information.
+
+.. code:: c++
+   
+    class FunctionPassNode : PassNode {
+      PassInfo pass_info;
+      runtime::TypedPackedFunc<Function(Function, Module, PassContext)> pass_func;
+      Module operator()(const Module& mod, const PassContext& pass_ctx) const final;
+      bool SkipFunction(const Function& func) const;
+      // Other members/methods are omitted...
+    };
+
+``pass_info`` is identical to what we just described in the module pass.
+``pass_func`` takes a function for optimization, it also needs a module as we
+may use it for reporting errors. A function could be annotated with
+"SkipOptimization" so that it will be ignored during optimization.
+
+Sequential Passes
+^^^^^^^^^^^^^^^^^
+
+SequentialPass is similar to Pytorch nn.Sequential that contains a host of
+passes for execution. It also has a disabled list that contains the list of
+disabled passes. For example, we may have a configuration which performs all
+passes at ``opt_level=3`` with some disabled passes using ``disabled_pass=xx``
+provided by ``PassContext``. Now using ``SequentialPass``, we could glob all
+passes at ``opt_level=3`` and exclude those in disabled.
+
+.. code:: c++
+
+    class SequentialPassNode : PassNode {
+      PassInfo pass_info;
+      // Passes need to be executed.
+      Array<Pass> passes;
+      // Passes are disabled during optimization
+      std::vector<std::string> disabled;
```

Comment:
Why do not you use Array<StringImm> ?

Reply:
Thanks for pointing out. `disabled` shouldn't belong here. We only have it in `PassContext` which is a `Array<tvm::Expr>`